slider/beatmap.py

```diff
@@ -909,6 +909,56 @@ def _get_as_bool(groups, section, field, default=no_default):
         )
 
 
+def _moving_average_by_time(times, data, delta, num):
+    """Take the moving average of some values and sample it at regular
+    frequencies.
+
+    Parameters
+    ----------
+    times : np.ndarray
+        The array of times to use in the average.
+    data : np.ndarray
+        The array of values to take the average of. Each column is averaged independently.
+    delta : int or float
+        The length of the leading and trailing window in seconds
+    num : int
+        The number of samples to take.
+
+    Returns
+    -------
+    times : np.ndarray
+        A column vector of the times sampled at.
+    averages : np.ndarray
+        A column array of the averages. 1 column per column in the input
+    """
+
+    # take an even sample from 0 to the end time
+    out_times = np.linspace(times[0].item(), times[-1].item(), num, dtype='timedelta64[ns]')
+    delta = np.timedelta64(int(delta * 1e9), 'ns')
+
+    # compute the start and stop indices for each sampled window
+    window_start_ixs = np.searchsorted(times[:, 0], out_times - delta)
+    window_stop_ixs = np.searchsorted(times[:, 0], out_times + delta)
+
+    # a 2d array of shape ``(num, 2)`` where each row holds the start and stop
+    # index for the window
+    window_ixs = np.stack([window_start_ixs, window_stop_ixs], axis=1)
+
+    # append a nan to the end of the values so that we can do many slices all
+    # the way to the end in reduceat
+    values = np.vstack([data, [np.nan] * data.shape[1]])
+
+    # sum the values in the ranges ``[window_start_ixs, window_stop_ixs)``
+    window_sums = np.add.reduceat(values, window_ixs.ravel())[::2]
+    window_sizes = np.diff(window_ixs, axis=1).ravel()
+    # convert window_sizes of 0 to 1 (inplace) to prevent division by zero
+    np.clip(window_sizes, 1, None, out=window_sizes)
+
+    out_values = np.stack(window_sums / window_sizes.reshape((-1, 1)))
+
+    return out_times.reshape((-1, 1)), out_values
```

Comment:
in general I prefer `[:, np.newaxis]` to `.reshape((-1, 1))`. imo it reads better.